### PR TITLE
refactor eslint configuration

### LINKED ISSUE #3245: Eslint hangs
Running `pnpm lint` won't finish.

image

**Environment**:

- OS:macOS 12

### PR DESCRIPTION
**Pull request by DukeManh**
I try my best to refactor our shared ESlint configuration. 

Changes made:
 - Since we publish `@senecacdot/eslint-config-telescope` as a package, other eslint configure can extend it just like `airbnb-config`, for example. Other configuration doesn't have to go into `overrides`
 - Use EcmaVersion 2021
 - Use recommended react-native config
 - Install SuperTest as main dependency instead of dev, following ESlint recommendation
 - Fix #3245 by changing eslint file pattern from `**/*` (which might include outside files) to `./**/*`
 - Add 'ts,tsx' to file import resolver
 - Use .gitignore as a .eslintignore (so cool)
 - Use '@typescript-eslint/eslint-plugin' in `src/web` only
 - Use '@eslint-plugin-react-native' in src/mobile only

### CODE REVIEW

File: tools/eslint/package.json [hunk 1]
```diff
@@ -5,6 +5,9 @@
   "main": "index.js",
   "repository": "Seneca-CDOT/telescope",
   "license": "BSD-2-Clause",
+  "peerDependencies": {
+    "eslint": "7.32.0"
+  },
```
Comment (cindyledev) [inline tools/eslint/package.json:10]: Wow, I didn't know there was a `peerDependencies` key we could use. I've only ever seen `devDependencies` and `dependencies`

File: src/api/dependency-discovery/package.json [hunk 2]
```diff
@@ -22,7 +22,8 @@
   "homepage": "https://github.com/Seneca-CDOT/telescope#readme",
   "dependencies": {
     "@senecacdot/satellite": "^1.27.0",
-    "query-registry": "2.2.0"
+    "query-registry": "2.2.0",
+    "supertest": "6.1.6"
```
Comment (cindyledev) [inline src/api/dependency-discovery/package.json:26]: What's this supertest dep used for?

**Comment by DukeManh** [comment]
The ESLint test is expected to fail because it's still using the old config.

File: src/api/dependency-discovery/package.json (hunk 2 above)
Comment (DukeManh) [reply src/api/dependency-discovery/package.json:26]: ESLint insists that I install it as and production dependency :). We are using it in our tests.

File: src/web/package.json [hunk 3]
```diff
@@ -38,7 +38,8 @@
     "react-use": "17.3.2",
     "smoothscroll-polyfill": "0.4.4",
     "swr": "1.2.2",
-    "yup": "0.32.11"
+    "yup": "0.32.11",
+    "@testing-library/react": "12.1.4"
```
Comment (cindyledev) [inline src/web/package.json:42]: Just a nit but `@` precedes alpha letters

File: src/web/package.json [hunk 4]
```diff
@@ -48,6 +49,8 @@
     "babel-loader": "8.2.3",
     "eslint": "7.32.0",
     "terser-webpack-plugin": "5.3.1",
-    "typescript": "4.4.4"
+    "typescript": "4.4.4",
+    "@typescript-eslint/eslint-plugin": "4.33.0",
+    "@typescript-eslint/parser": "4.33.0"
```
Comment (cindyledev) [inline src/web/package.json:54]: Same here

File: src/api/dependency-discovery/package.json (hunk 2 above)
Comment (cindyledev) [reply src/api/dependency-discovery/package.json:26]: Ah okay

File: src/web/package.json [hunk 5]
```diff
@@ -38,7 +38,8 @@
     "react-use": "17.3.2",
     "smoothscroll-polyfill": "0.4.4",
     "swr": "1.2.2",
-    "yup": "0.32.11"
+    "yup": "0.32.11",
+    "@testing-library/react": "12.1.4"
   },
   "devDependencies": {
     "@senecacdot/eslint-config-telescope": "1.0.1",
```
Comment (cindyledev) [inline src/web/package.json:45]: The ESLint is failing but it's expected due to the `tools/eslint/index.js` changes in this PR
https://github.com/Seneca-CDOT/telescope/blob/4d411c6f58c1bac42fc14264aed9d33cc0feb4fd/tools/eslint/index.js#L14-L16

This `import/resolver` will resolve the 155 errors. To pass ESLint locally, change this line to
```
"@senecacdot/eslint-config-telescope": "workspace:1.0.1",
```

Comment (DukeManh) [reply src/web/package.json:45]: Thanks @cindy.
Or you can manually link eslint: `cd <subdir> && pnpm link ./<relativepath-to-eslint>

Comment (cindyledev) [reply src/web/package.json:45]: Err you tagged angry-raccoon-python-developer-cindy. Also good to know the pnpm link method. Knowing CLI commands makes things shorter :)

Comment (DukeManh) [reply src/web/package.json:45]: lol, those names are always taken

Comment (cindyledev) [reply src/web/package.json:45]: Unfortunately, I don't have an SEO-friendly name. Maybe when I'm married

**Review by cindyledev — APPROVED** [review]

**Comment by JiaHua-Zou** [comment]
I try rerun our CI. It looks like the eslint is failing.

**Comment by DukeManh** [comment]
> I tried rerunning the CI. It looks like the `eslint` is failing.

See @cindyledev's comment above

**Comment by DukeManh** [comment]
Rebased.

**Review by cindyledev — DISMISSED** [review]

**Review by humphd — CHANGES_REQUESTED** [review]
eslint in CI is unhappy

**Comment by DukeManh** [comment]
@humphd, because it's using Eslint-config@1.0.1, once this goes in we'll release a new version.

**Review by cindyledev — APPROVED** [review]

**Review by menghif — COMMENTED** [review]
You can increase the version of `eslint-config-telescope` here. I don't think we need to do it in a separate PR

**Comment by DukeManh** [comment]
What's the appropriate version of the package do you think.

**Comment by humphd** [comment]
`1.1.0` of `1.0.2`, either is fine.

**Comment by DukeManh** [comment]
Ok I updated the version, I assume it will be automatically published? Then, we can trigger a Renovate bot for bump v1.0.1 -> v1.1.0 in sub-packages.

**Comment by humphd** [comment]
Correct.  Merging on master will update it.

**Review by cindyledev — APPROVED** [review]

**Review by joelazwar — APPROVED** [review]

**Review by humphd — APPROVED** [review]

**Pull request merged**